src/Ionide.ProjInfo/ProjectLoader2.fs

```diff
@@ -0,0 +1,559 @@
+        result.ResultsByNode
+        |> Seq.map (fun (KeyValue(k, v)) ->
+            match v.OverallResult with
+            | BuildResultCode.Success -> KeyValuePair(k, Ok v)
+            | _ ->
+                let logs =
+                    errorLogs
+                    |> List.filter (fun e -> e.ProjectFile = k.ProjectInstance.FullPath)
+
+                KeyValuePair(k, Error('e.BuildFailure(v, logs)))
+        )
+        |> Dictionary<_, _>
+
+    let isolateFailures (result: GraphBuildResult, errorLogs: BuildErrorEventArgs list) =
+        resultsByNode (result, errorLogs)
+        |> Seq.choose (fun (KeyValue(k, v)) ->
+            match v with
+            | Ok v -> None
+            | Error e -> Some(k, e)
+        )
+
+/// <summary>
+/// Uses <see cref="T:Microsoft.Build.Execution.BuildManager"/> to run builds.
+/// This should be treated as a singleton because the BuildManager only allows one build request running at a time.
+/// </summary>
+type BuildManagerSession(?bm: BuildManager) =
+    let locker = BuildManagerSession.locker
+    let bm = defaultArg bm BuildManager.DefaultBuildManager
+
+    let tryGetErrorLogs (buildParameters: BuildParameters) =
+        buildParameters.Loggers
+        |> Seq.tryPick (
+            function
+            | :? ProjectLoader.ErrorLogger as e -> Some e
+            | _ -> None
+        )
+        |> Option.toList
+        |> Seq.collect (fun e -> e.Errors)
+        |> Seq.toList
+
+    let lockAndStartBuild (ct: CancellationToken) buildParameters (a: unit -> Task<_>) =
+        task {
+            use! _lock = locker.LockAsync ct
+            use _ = bm.StartBuild(buildParameters, ct)
+            return! a ()
+        }
+
+    member private x.determineBuildOutput<'e when BuildResultFailure<'e>>(buildParameters, result: BuildResult) =
+        match result.OverallResult with
+        | BuildResultCode.Success -> Ok result
+        | _ -> Error('e.BuildFailure(result, tryGetErrorLogs buildParameters))
+
+
+    member private x.determineGraphBuildOutput<'e when GraphBuildResultFailure<'e>>(buildParameters, result: GraphBuildResult) =
+        match result.OverallResult with
+        | BuildResultCode.Success -> Ok result
+        | _ -> Error('e.BuildFailure(result, tryGetErrorLogs buildParameters))
+
+
+    /// <summary>Submits a graph build request to the current build and starts it asynchronously.</summary>
+    /// <param name="buildRequest">GraphBuildRequestData encapsulates all of the data needed to submit a graph build request.</param>
+    /// <param name="buildParameters">All of the settings which must be specified to start a build</param>
+    /// <param name="ct">CancellationToken to cancel build submissions.</param>
+    /// <returns>The BuildResult</returns>
+    member x.BuildAsync(buildRequest: BuildRequestData, ?buildParameters: BuildParameters, ?ct: CancellationToken) =
```

Comment:
There's a recent PR against MSbuild to add this as a core capability: https://github.com/dotnet/msbuild/pull/12270

It'll be a bit until it lands and then is usable, but maybe the implementation can serve as a reference in the meantime.